File: lib/downloads.ts
// Download limits. max_downloads is null for an unlimited share; otherwise the
// share is deleted after that many downloads. The server does the atomic count
// and delete in store.RegisterDownload.

const MAX_CAP = 10_000;

/** Remaining downloads, or null when unlimited. */
export function downloadsLeft(count: number, max: number | null): number | null {
  if (max === null) return null;
  return Math.max(0, max - count);
}

export function isExhausted(count: number, max: number | null): boolean {
  return max !== null && count >= max;
}

/**
 * Clamps an uploader-supplied limit to 1..MAX_CAP, or returns null, meaning
 * unlimited, for a missing, zero, negative or fractional value.
 */
export function parseMaxDownloads(
  input: number | null | undefined,
): number | null {
  if (input == null || !Number.isInteger(input) || input < 1) return null;
  return Math.min(input, MAX_CAP);
}

File: server-go/internal/share/downloads.go
package share

// The download limit helpers follow lib/downloads.ts. A nil max is an unlimited
// share; store.RegisterDownload does the atomic count and burn.

const maxDownloadsCap = int64(10_000)

// DownloadsLeft returns the remaining downloads, never negative, or nil for an
// unlimited share.
func DownloadsLeft(count int64, max *int64) *int64 {
	if max == nil {
		return nil
	}
	left := *max - count
	if left < 0 {
		left = 0
	}
	return &left
}

// IsExhausted reports whether a limited share has used up its downloads.
func IsExhausted(count int64, max *int64) bool {
	return max != nil && count >= *max
}

// ParseMaxDownloads clamps an uploader-supplied limit to 1..maxDownloadsCap and
// returns nil, meaning unlimited, for a missing, zero or negative value.
func ParseMaxDownloads(input *int64) *int64 {
	if input == nil || *input < 1 {
		return nil
	}
	v := *input
	if v > maxDownloadsCap {
		v = maxDownloadsCap
	}
	return &v
}
